cpp/src/hash/hashing.cu

```diff
@@ -417,7 +438,7 @@ hash_partition_table(table_view const& input,
 {
   auto const num_rows = table_to_hash.num_rows();
 
-  bool const use_optimization {num_partitions <= 512};
+  bool const use_optimization {num_partitions <= 1024};
```

Comment:
What determines the value 1024? Is it 2x the block size (512)? If so, use that rather than just the number. Otherwise, at least document the intent.

Reply:
The optimized code is limited by shared memory size and when `num_partitions ` is large (>1024 partitions) we won't get enough shared memory according to @trevorsm7 's tests. In which case, we simply switch back to the scatter approach. That's why we only apply the optimization when `num_partitions <= 1024`. I added a constant called `THRESHOLD_FOR_OPTIMIZED_PARTITION_KERNEL` for this value (1024).